src/ui/actions/timeline.ts

```diff
@@ -504,50 +492,77 @@ export function setPlaybackPrecachedTime(time: number): SetPlaybackPrecachedTime
   return { type: "set_playback_precached_time", time };
 }
 
-export function setFocusRegion(focusRegion: {
-  startTime: number;
-  endTime: number;
-}): SetFocusRegionAction {
-  return { type: "set_trim_region", focusRegion };
-}
-
-export function setFocusAroundTime(time: number, maxTime: number): UIThunkAction {
+export function setFocusRegion(focusRegion: FocusRegion | null): UIThunkAction {
   return (dispatch, getState) => {
     const state = getState();
-    const zoomRegion = getZoomRegion(state);
+    const currentTime = getCurrentTime(state);
 
-    const duration = zoomRegion.endTime;
-    const startTime = Math.max(time - maxTime / 2, 0);
-    const endTime = Math.min(time + maxTime / 2, duration);
+    // Stop playback (if we're playing) to avoid the currentTime from getting out of bounds.
+    const playback = getPlayback(state);
+    if (playback !== null) {
+      dispatch(stopPlayback());
+    }
 
-    dispatch(setFocusRegion({ endTime, startTime }));
-  };
-}
+    if (focusRegion !== null) {
+      const zoomRegion = getZoomRegion(state);
+      const { endTime: prevEndTime, startTime: prevStartTime } = getFocusRegion(state) || {};
 
-export function updateFocusRegion(operation: FocusOperation): UIThunkAction {
-  return (dispatch, getState) => {
-    const state = getState();
-    const hoverTime = getHoverTime(state)!;
-    const focusRegion = getFocusRegion(state)!;
-    const zoomRegion = getZoomRegion(state);
+      let { endTime, startTime } = focusRegion;
 
-    const duration = zoomRegion.endTime;
-    // To get around a handful of selection UI bugs, this limits the allowable trimmed
-    // region to 10% of the full duration of the replay.
-    const minRegion = duration * 0.1;
-    const { startTime, endTime } = focusRegion;
+      // Basic bounds check.
+      if (startTime < zoomRegion.startTime) {
+        startTime = zoomRegion.startTime;
+      }
+      if (endTime > zoomRegion.endTime) {
+        endTime = zoomRegion.endTime;
+      }
 
-    let value, type;
+      // Make sure our region is valid.
+      if (endTime < startTime) {
+        // If we need to adjust a dimension, it's the most intuitive to adjust the one that's being updated.
+        if (prevEndTime === endTime) {
+          startTime = endTime;
+        } else {
+          endTime = startTime;
+        }
+      }
+
+      // Make sure the current time stays within the bounds of our selected region.
+      if (currentTime < startTime) {
+        dispatch(setTimelineState({ currentTime: startTime }));
+      } else if (currentTime > endTime) {
+        dispatch(setTimelineState({ currentTime: endTime }));
```

Comment:
Right, the naming of the actions in this bit of state is pretty confusing. (I didn't create them. I actually deleted a few in this PR.)

`setTimelineState` is kind of a cheat for updating any part of the state:
https://github.com/RecordReplay/devtools/blob/d08cd335a3435baf28f0cb84566070c3a121d681/src/ui/reducers/timeline.ts#L33-L35

I think this kind of defeats the purpose of the action/reducer pairing. But let's leave it for now since it's used in a few places and we can clean this up in a follow up.